pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java

```diff
@@ -308,39 +348,44 @@ String getHandlerName() {
-                        .map(partitionName -> {
-                            int partitionIndex = TopicName.getPartitionIndex(partitionName);
-                            ProducerImpl<T> producer =
-                                new ProducerImpl<>(client,
-                                    partitionName, conf, new CompletableFuture<>(),
-                                    partitionIndex, schema, interceptors);
-                            producers.add(producer);
-                            return producer.producerCreatedFuture();
-                        }).collect(Collectors.toList());
-
-                    FutureUtil.waitForAll(futureList)
-                        .thenAccept(finalFuture -> {
-                            if (log.isDebugEnabled()) {
-                                log.debug("[{}] success create producers for extended partitions. old: {}, new: {}",
-                                    topic, oldPartitionNumber, currentPartitionNumber);
-                            }
-                            topicMetadata = new TopicMetadataImpl(currentPartitionNumber);
-                            future.complete(null);
-                        })
-                        .exceptionally(ex -> {
-                            // error happened, remove
-                            log.warn("[{}] fail create producers for extended partitions. old: {}, new: {}",
-                                topic, oldPartitionNumber, currentPartitionNumber);
-                            List<ProducerImpl<T>> sublist = producers.subList(oldPartitionNumber, producers.size());
-                            sublist.forEach(newProducer -> newProducer.closeAsync());
-                            sublist.clear();
-                            future.completeExceptionally(ex);
-                            return null;
-                        });
-                    return null;
+                    if (conf.isLazyStartPartitionedProducers() && conf.getAccessMode() == ProducerAccessMode.Shared) {
+                        topicMetadata = new TopicMetadataImpl(currentPartitionNumber);
+                        future.complete(null);
+                        return future;
+                    } else {
+                        List<CompletableFuture<Producer<T>>> futureList = list
+                                .subList(oldPartitionNumber, currentPartitionNumber)
+                                .stream()
+                                .map(partitionName -> {
+                                    int partitionIndex = TopicName.getPartitionIndex(partitionName);
+                                    ProducerImpl<T> producer =
+                                            new ProducerImpl<>(client,
+                                                    partitionName, conf, new CompletableFuture<>(),
+                                                    partitionIndex, schema, interceptors);
+                                    producers.put(partitionIndex, producer);
```

Comment:
Wouldn't this be replaced with `computeIfAbsent`? For now, there seems to be no conflict between here and other processes. But that may not be the case in the future.

Reply:
I've replaced to use `computeIfAbsent`.